File: 백준/Gold/14502. 연구소/연구소.java

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int n, m;
    static int[][] matrix;
    static int[][] copyMatrix;
    static int answer = 0;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        matrix = new int[n][m];
        copyMatrix = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                int x = Integer.parseInt(st.nextToken());
                matrix[i][j] = x;
            }
        }

//        for (int i = 0; i < n; i++) {
//            System.out.println(Arrays.toString(matrix[i]));
//        }

        buildWalls(0);
        System.out.println(answer);

    }

    private static void buildWalls(int count) {

        if (count == 3) {
            copyMatrix = copyMatrix();

            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (copyMatrix[i][j] == 2) {
                        spreadVirus(i,j);
                    }
                }
            }
            answer = Math.max(answer, getSafetyArea());
            return;
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][j] = 1;
                    buildWalls(count+1);
                    matrix[i][j] = 0;
                }
            }
        }
    }

    // 벽이 세워진 메트릭스 복사
    private static int[][] copyMatrix() {
        int[][] arr = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                arr[i][j] = matrix[i][j];
            }
        }
        return arr;
    }

    private static void spreadVirus(int row, int col) {
        int[][] directions = {
                {-1,0}, {1,0}, {0,-1}, {0,1}
        };

        for(int[] direction : directions) {
            int newRow = row + direction[0];
            int newCol = col + direction[1];

            if (newRow>=0 && newRow<n && newCol>=0 && newCol<m){
                if (copyMatrix[newRow][newCol] == 0) {
                    copyMatrix[newRow][newCol] = 2;
                    spreadVirus(newRow, newCol);
                }
            }
        }
    }

    private static int getSafetyArea() {
        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if(copyMatrix[i][j] == 0) {
                    result++;
                }
            }
        }
        return result;
    }
}

File: 백준/Gold/14502. 연구소/연구소.py

# 입력
n, m = map(int, input().split(" "))
matrix = []
temp = [[0]*m for _ in range(n)]


for _ in range(n):
    matrix.append(list(map(int, input().split(" "))))

result = -1

def build_walls(count):
    global result

    if count == 3:
        copy_matrix()

        check_virus()
        result = max(result, get_safety_area())
        return

    for i in range(n):
        for j in range(m):
            if matrix[i][j] == 0:
                matrix[i][j] = 1
                build_walls(count + 1)
                matrix[i][j] = 0


def copy_matrix():
    for i in range(n):
        for j in range(m):
            temp[i][j] = matrix[i][j]


def check_virus():
    for i in range(n):
        for j in range(m):
            if temp[i][j] == 2:
                spread_virus(i, j)


def spread_virus(row, col):
    d_row = [-1, 1, 0, 0]
    d_col = [0, 0, -1, 1]

    for i in range(4):
        new_row = row + d_row[i]
        new_col = col + d_col[i]

        if 0 <= new_row < n and 0 <= new_col < m:
            if temp[new_row][new_col] == 0:
                temp[new_row][new_col] = 2
                spread_virus(new_row, new_col)

                
def get_safety_area():
    res = 0

    for i in range(n):
        for j in range(m):
            if temp[i][j] == 0:
                res += 1
    return res

# 실행
build_walls(0)
print(result)